### PR TITLE
Addition of long-reads pre-assembly qcs

### PR DESCRIPTION
**Pull request by Ge94**
Hi! This is the first PR for the LR assembly pipeline. As of now I have taken care of general structure, rough integration with the miassembler, ONT pre-qcs and I am missing only one module to complete pacbio's pre-qcs. I also installed the modules that I should use for the next steps. As previously mentioned, I am not looking for a deep review yet, but if you see anything wrong let me know please! 

The plan for the upcoming steps is to have four tracks to assemble different types of data: low-quality ONT, high-quality ONT, low-quality pacbio and pacbio-hifi. The structure will follow the four columns illustrated [here](https://app.diagrams.net/#G1RlYB8ZjwwliVjx5MEbcmeV3pfrM5Er6l#%7B%22pageId%22%3A%22jNuOrICNevvYicqJh7_j%22%7D).

Right now there are two separate workflows in the "workflows" folder. It would be nice to get some more integration covered in the internal nf hackathon. 

The only module I updated, rather than entirely downloaded with nf-core, is fastp. I tried integrating the two the best way I could (the pipelines were using different nf-core versions). I know for sure I haven't updated the fastp.diff file, how could that be done?

Thank you in advance!

### CODE REVIEW

**Comment by mberacochea** [comment]
> The only module I updated, rather than entirely downloaded with nf-core, is fastp. I tried integrating the two the best way I could (the pipelines were using different nf-core versions). I know for sure I haven't updated the fastp.diff file, how could that be done?

Using the nf-core utilities: `nf-core modules patch fastp` or similar

File: subworkflows/local/long_reads_qc.nf [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        [],
+        false,
+        false,
+        false,
+        false
+    )
+
+    ch_versions = ch_versions.mix(FASTP_LR.out.versions)
+
+    RAW_READ_QUALITY_CHECK(
```
Comment (mberacochea) [inline subworkflows/local/long_reads_qc.nf:26]: I would remove this one with in favour of json splirping and the branch operator.

```groovy
quality_levels_ch = FASTP_LR.out.json.map { meta, json -> {
        json_txt = new JsonSlurper().parseText(json.text)
        q20bases = json_txt?.summary?.before_filtering?.q20_bases ?: 0;
        total_bases = json_txt?.summary?.before_filtering?.total_bases ?: 0;

        q20_percentage = q20_bases / total_bases * 100

        categorization = [
            "high_quality": q20_percentage >= 80, 
            "low_quality": q20_percentage < 80,
        ]
        return [meta, categorization]
    } 
}
```
It's also worth discarding data that was heavily filtered, like we do for short reads.

File: subworkflows/local/long_reads_qc.nf [hunk 2]
```diff
@@ -0,0 +1,95 @@
+include { FASTP as FASTP_LR                       } from '../../modules/nf-core/fastp/main'
+include { RAW_READ_QUALITY_CHECK                  } from '../../modules/local/raw_read_quality_check/'
+include { MINIMAP2_ALIGN as HUMAN_DECONTAMINATION } from '../../modules/nf-core/minimap2/align/main'
```
Comment (mberacochea) [inline subworkflows/local/long_reads_qc.nf:3]:
```suggestion
include { MINIMAP2_ALIGN as MINIMAP2_ALIGN_HUMAN } from '../../modules/nf-core/minimap2/align/main'
```

File: subworkflows/local/long_reads_qc.nf [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+include { MINIMAP2_ALIGN as HOST_DECONTAMINATION  } from '../../modules/nf-core/minimap2/align/main'
```
Comment (mberacochea) [inline subworkflows/local/long_reads_qc.nf:4]:
```suggestion
include { MINIMAP2_ALIGN as MINIMAP2_ALIGN_HOST  } from '../../modules/nf-core/minimap2/align/main'
```

File: subworkflows/local/long_reads_qc.nf [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                files -> [ ["id": params.human_fasta_prefix], files ]
+            }
+
+        // TODO: can we change the way human/host are given via prefixes?
+
+        HUMAN_DECONTAMINATION(
+            FASTP_LR.out.reads,
+            human_reference,
+            "human",
+            true,
```
Comment (mberacochea) [inline subworkflows/local/long_reads_qc.nf:48]: Can you add comments on the parameters this is recieving?

For example:
```
HUMAN_DECONTAMINATION(
            FASTP_LR.out.reads,
            human_reference,
            "human",
            true, # Filter out
            "bai", # Output format...
            false,
            true
        )
```

Stuff like that, it makes reading the code much easier when multiples parameters are send to a module like this

File: subworkflows/local/ont_hq.nf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+include { PORECHOP_ABI as PORECHOP_ONT         } from '../../modules/nf-core/porechop/abi/main'
```
Comment (mberacochea) [inline subworkflows/local/ont_hq.nf:1]: No need to rename the module here, keep the original name please

Comment (mberacochea) [reply subworkflows/local/ont_hq.nf:1]: Also, no need for this to be a subworkflow if it's only one module

File: subworkflows/local/ont_lq.nf [hunk 6]
```diff
@@ -0,0 +1,18 @@
+include { CANU as CANU_ONT                     } from '../../modules/nf-core/canu/main'
```
Comment (mberacochea) [inline subworkflows/local/ont_lq.nf:1]: Same comments as ont_hq swf

**Review by mberacochea — COMMENTED** [review]
Good stuff @Ge94 , I've left a few comments :)

File: subworkflows/local/ont_hq.nf (hunk 5 above)
Comment (Ge94) [reply subworkflows/local/ont_hq.nf:1]: I agree on the module name. The rest of the workflow is WIP, I just wrote the main structure in this version

File: modules/local/fetchtool_reads.nf [hunk 7]
```diff
@@ -32,6 +32,15 @@ process FETCHTOOL_READS {
     library_strategy=\$(echo "\$(grep ${reads_accession} download_folder/${study_accession}/${study_accession}.txt | cut -f 7)" | tr '[:upper:]' '[:lower:]')
     library_layout=\$(echo "\$(grep ${reads_accession} download_folder/${study_accession}/${study_accession}.txt | cut -f 5)" | tr '[:upper:]' '[:lower:]')
 
+    export metadata_platform=\$(echo "\$(grep ${reads_accession} download_folder/${study_accession}/${study_accession}.txt | cut -f 8)" | tr '[:upper:]' '[:lower:]')
+    if [[ \$metadata_platform == "minion" || \$metadata_platform == "promethion" || \$metadata_platform == "gridion" ]]; then
+        platform="ont"
+    elif [[ \$metadata_platform == "pacbio rs" || \$metadata_platform == "pacbio rs ii" ]]; then
+        platform="pacbio"
+    else
+        platform=\$metadata_platform
```
Comment (mberacochea) [inline modules/local/fetchtool_reads.nf:41]:
```suggestion
        platform="\$metadata_platform"
```

File: bin/check_raw_quality.py
Comment (mberacochea) [inline bin/check_raw_quality.py:1]: I would delete this file

**Review by mberacochea — COMMENTED** [review]
2 more comments :)

File: modules/local/fetchtool_reads.nf (hunk 7 above)
Comment (mberacochea) [reply modules/local/fetchtool_reads.nf:41]: In case the platform consists of multiple words.

File: conf/test.config [hunk 8]
```diff
@@ -21,6 +21,14 @@ profiles {
             bwamem2_reference_genomes_folder = "${projectDir}/tests/human_phix/bwa2mem"
             blast_reference_genomes_folder   = "${projectDir}/tests/human_phix/blast"
             reference_genomes_folder         = "${projectDir}/tests/human/"
+
+            max_spades_retries               = -1
+            max_megahit_retries              = -1
```
Comment (mberacochea) [inline conf/test.config:26]: image

**Review by mberacochea — APPROVED** [review]
One tiny comment.. I will test the tests before merging if you don't mind :)

**Pull request merged**